blog/2021/08/DEQ.md

```diff
@@ -0,0 +1,426 @@
+As it turns out, there is a line of work following [Deep Equilibrium Models](https://arxiv.org/abs/1909.01377)
+that explores the notion of infinitely-deep neural networks in an explicit manner. It borrows
+much of the ideas from the NeuralODE line of research, and allows neural network structures
+to be viewed as repeated applications of layered computations. This is often very accurate,
+for example when we apply convolution filters on images---the network can essentially
+be viewed as a convolutional feature extractor consisting of repetitive blocks, and one
+linear output layer at the very end. When there is an infinite number of these repeated
+blocks, the network might explode off to infinity. If it doesn't, however, it reaches an
+equilibrium points and converges, hence the name.
+
+## DEQ Models and Infinitely Deep Networks
+
+Let us begin by looking at an instantiation of the equations for DEQ models.
+$$
+\begin{array}{l}
+z:=0\\
+\text { Repeat until convergence: }\\
+z:=\tanh (W z+x)
+\end{array}
+$$
+
+Which eventually brings us to
+$$
+z^{\star}=\tanh \left(W z^{\star}+x\right)
+$$
+
+The idea of DEQ is, in a sense, insanely simple! We assume there are infinitely many
+compositions of identical layers (equivalently an infinitely deep network), and
+directly solve for the fixed point described by this equation. Clearly, naively iterating
+for the fixed point would take an enormous amount of time. Thanks to the Implicit Function
+Theorem, DEQ shows that it is possible to train an infinitely-deep neural network using the
+gradient optimization toolbox. For more reference, please see
+[Deep Implicit Layers](http://implicit-layers-tutorial.org)
+
+From the viewpoint of Julia and the DiffEqFlux.jl library, it is also natural to look at DEQ from
+a differential equations perspective. Thanks to NeuralODE, the machine learning community is now able
+to put on the glasses of differential equations, and see through a new perspective.
+If we perceive the iterating layer of the network as an ODE, DEQ is solving nothing but the
+well-known "Steady State Problem", defined as the final converging state of the ODE as time
+goes to infinity. This line of work is well-studied in the scientific machine learning
+literature, see for example [Notes on Adjoint Methods for 18.335](https://math.mit.edu/~stevenj/18.336/adjoint.pdf)
+
+$$
+f(u) - u = du = 0
+$$
+
+Reaching a fixed point of the iterative layer application is much like solving for the final
+value in a converging sequence. To put it simply, we set the current value equal to the next,
+and obtain a solution for this recursion problem. As we show above, if we use `u` to represent
+the current layer and `f(u)` to represent the next layer, we can sloppily denote the fixed point
+equation as `f(u)-u=0`. Crucially, this can also be converted into a steady state problem
+by defining an ODE layer `f(u)-u=du` and solving
+that ODE in its steady state. This gives us equality among all three components.
```

Comment:
```suggestion
## Mixing DEQs and Neural ODEs

From the viewpoint of Julia and the DiffEqFlux.jl library, it is also natural to look at DEQ from
a differential equations perspective. Thanks to NeuralODE, the machine learning community is now able
to put on the glasses of differential equations, and see through a new perspective. Instead of viewing the
dynamical system as a discrete process $x_{n+1} = x_n + f(x_n)$, we can equivalently view the system as evolving
continuously, i.e. $x' = f(x)$. In this sense, convergence is when the change is zero, or $x'=0$, which again
happens when $f(x)=0$ and is a rootfinding problem. But this view is insightful: a DEQ is a neural ODE where
time goes to infinity.
```